// @embrace-io/web-sdk/build/src/resources/constants/index.js
'use strict';

// This will be replaced after bundling the app by the Embrace SourceUpload CLI tool.
// It needs to always remain 32 chars long, so we don't mess with the sourcemaps location after replacing the bundle id with the real 32 chars long bundle ID
const TEMPLATE_BUNDLE_ID = 'EmbIOBundleIDfd6996f1007b363f87a';
const TEMPLATE_APP_VERSION = 'EmbIOAppVersionX.X.X';
const SDK_VERSION = '1.8.1';
const EMBRACE_SERVICE_NAME = 'embrace-web-sdk';
// NATIVE_FRAMEWORK is the representation Embrace BE uses to differentiate between
// sdks. It is an enum, so we just send a number here
const NATIVE_FRAMEWORK = 1;
const EMBRACE_APP_INSTANCE_ID_STORAGE_KEY = 'embrace_app_instance_id';

exports.EMBRACE_APP_INSTANCE_ID_STORAGE_KEY = EMBRACE_APP_INSTANCE_ID_STORAGE_KEY;
exports.EMBRACE_SERVICE_NAME = EMBRACE_SERVICE_NAME;
exports.NATIVE_FRAMEWORK = NATIVE_FRAMEWORK;
exports.SDK_VERSION = SDK_VERSION;
exports.TEMPLATE_APP_VERSION = TEMPLATE_APP_VERSION;
exports.TEMPLATE_BUNDLE_ID = TEMPLATE_BUNDLE_ID;
